modules/gmake/src/proj.lua

```diff
@@ -0,0 +1,677 @@
+	else
+		wl('ALL_CXXFLAGS += $(CXXFLAGS) $(ALL_CPPFLAGS)')
+	end
+end
+
+
+---
+-- Prints the linker flags for the given project.
+--
+-- @param prj
+--  project to print linker flags for
+---
+function proj.linkFlags(prj)
+	local toolset = prj.linker
+	local flags = toolset.getLinkerFlags(prj)
+
+	if #flags > 0 then
+		wl('ALL_LDFLAGS = $(LDFLAGS) %s', table.concat(flags, ' '))
+	else
+		wl('ALL_LDFLAGS = $(LDFLAGS)')
+	end
+end
+
+
+---
+-- Prints all configurations for the project.
+--
+-- @param prj
+--  project to print configurations for
+---
+function proj.configurations(prj)
+	wl('# Configuration-level overrides')
+	for _, cfg in ipairs(prj.configs) do
+		local name = cfg.name:lower()
+
+		local conditional = nil
+		if _ == 1 then
+			conditional = 'ifeq ($(config), ' .. name .. ')'
+		else
+			conditional = 'else ifeq ($(config), ' .. name .. ')'
+		end
+
+		wl(conditional)
+
+		export.indent()
+		premake.callArray(proj.elements.configurations, cfg)
+		export.outdent()
+	end
+
+	if #prj.configs > 0 then
+		wl('endif')
+	end
+
+	wl()
+end
+
+
+---
+-- Prints the target directory for the given configuration.
+--
+-- @param cfg
+--  configuration to print target directory of
+---
+function proj.configTargetDir(cfg)
+	-- TODO: Check if this there is an override of the default dir before writing the TARGETDIR
+	local prj = cfg.project
+	local cfgName = cfg.name
+	local configTargetDir = path.getRelative(prj.location, path.join(prj.workspace.location, 'bin', prj.name, cfgName))
+	wl('TARGETDIR = %s', configTargetDir)
+end
+
+
+---
+-- Prints the intermediates directory for the given configuration.
+--
+-- @param cfg
+--  configuration to print intermediates directory of
+---
+function proj.configIntermediateDir(cfg)
+	-- TODO: Check if this there is an override of the default dir before writing the OBJDIR
+	local prj = cfg.project
+	local cfgName = cfg.name
+	local intDir = path.getRelative(prj.location, path.join(prj.workspace.location, 'obj', prj.name, cfgName))
+	wl('OBJDIR = %s', intDir)
+end
+
+
+---
+-- Prints the target name for the given configuration.
+--
+-- @param cfg
+--  configuration to print target name of
+---
+function proj.configTargetName(cfg)
+	--- TODO: Override target name
+end
+
+
+---
+-- Prints the defines for the given configuration.
+--
+-- @param cfg
+--  configuration to print defines of
+---
+function proj.configDefines(cfg)
+	local defs = cfg.defines
+	if defs ~= nil and #defs > 0 then
+		local defString = table.concat(table.map(defs, function(key, value)
+			return '-D' .. value
+		end), ' ')
+		wl('DEFINES += %s', defString)
+	end
+end
+
+
+---
+-- Prints the C flags for the given configuration.
+--
+-- @param cfg
+--  configuration to print C flags of
+---
+function proj.configCFlags(cfg)
+	local toolset = cfg.compiler
+	local flags = toolset.getCFlags(cfg)
+
+	if #flags > 0 then
+		wl('ALL_CFLAGS += %s', table.concat(flags, ' '))
+	end
+end
```

Comment:
As above, would like to keep a single copy of this logic and use it for both `prj` and `cfg` when we can.